aptos-move/aptos-replay-benchmark/src/diff.rs

```diff
@@ -0,0 +1,223 @@
+        right: u64,
+    },
+    ExecutionStatus {
+        left: ExecutionStatus,
+        right: ExecutionStatus,
+    },
+    Event {
+        left: Option<ContractEvent>,
+        right: Option<ContractEvent>,
+    },
+    WriteSet {
+        left: Option<(StateKey, WriteOp)>,
+        right: Option<(StateKey, WriteOp)>,
+    },
+}
+
+/// Holds all differences for a pair of transaction outputs.
+pub(crate) struct TransactionDiff {
+    diffs: Vec<Diff>,
+}
+
+impl TransactionDiff {
+    /// Returns an empty diff - there are no differences between transaction outputs.
+    pub(crate) fn empty() -> Self {
+        Self { diffs: vec![] }
+    }
+
+    /// Given a pair of transaction outputs, computes its [TransactionDiff] that includes the gas
+    /// used, execution status, events and write sets.
+    // TODO: Make comparison configurable, so we can skip gas differences, etc.
+    pub(crate) fn from_outputs(left: TransactionOutput, right: TransactionOutput) -> Self {
+        let (left_write_set, left_events, left_gas_used, left_transaction_status, _) =
+            left.unpack();
+        let (right_write_set, right_events, right_gas_used, right_transaction_status, _) =
+            right.unpack();
+
+        let mut diffs = vec![];
+
+        // All statuses must be kept, since we are replaying transactions.
+        let left_execution_status = assert_ok!(left_transaction_status.as_kept_status());
+        let right_execution_status = assert_ok!(right_transaction_status.as_kept_status());
+        if left_execution_status != right_execution_status {
+            diffs.push(Diff::ExecutionStatus {
+                left: left_execution_status,
+                right: right_execution_status,
+            });
+        }
+
+        if left_gas_used != right_gas_used {
+            diffs.push(Diff::GasUsed {
+                left: left_gas_used,
+                right: right_gas_used,
+            });
+        }
+
+        Self::diff_events(&mut diffs, left_events, right_events);
+        Self::diff_write_sets(&mut diffs, left_write_set, right_write_set);
+
+        Self { diffs }
+    }
+
+    /// Returns true if the diff is empty, and transaction outputs match.
+    pub(crate) fn is_empty(&self) -> bool {
+        self.diffs.is_empty()
+    }
+
+    /// Computes the differences between a pair of event vectors, and adds them to the diff.
+    fn diff_events(diffs: &mut Vec<Diff>, left: Vec<ContractEvent>, right: Vec<ContractEvent>) {
+        let event_vec_to_map = |events: Vec<ContractEvent>| {
+            let mut ty_tagged_events = BTreeMap::new();
+            for event in events {
+                ty_tagged_events.insert(event.type_tag().clone(), event);
+            }
+            ty_tagged_events
+        };
+
+        let left = event_vec_to_map(left);
+        let mut right = event_vec_to_map(right);
+
+        for (left_ty_tag, left_event) in left {
+            if let Some(right_event) = right.remove(&left_ty_tag) {
```

Comment:
```
diffs.push(Diff::Event {
   left: Some(left_event),
   right: right.remove(&left_ty_tag).filter(|event| (left_event.event_data() != event.event_data())
}
```

Reply:
Actually this does not work, right? You do not want to get it mapped to None if event data is equal. So I changed to `is_sme_and` to skip instead.

Reply:
oh, not me here suggesting bugs :D

Reply:
Suggesting tests helps :D